docs/style-spec/expressions.md

```diff
@@ -0,0 +1,137 @@
+ - `interval` functions return the output value of the stop just less than the function input. The domain must be numeric. This is the default for properties marked with , the "interval" symbol.
+
+
+### Example: a zoom-only function.
+
+```js
+{
+  "circle-radius": {
+    "stops": [
+
+      // zoom is 5 -> circle radius will be 1px
+      [5, 1],
+
+      // zoom is 10 -> circle radius will be 2px
+      [10, 2]
+
+    ]
+  }
+}
+```
+
+### Example: a zoom-and-property function
+
+Using property functions as the output value for one or more zoom stops allows 
+the appearance of a map feature to change with both the zoom level _and_ the
+feature's properties.
+
+```js
+{
+  "circle-radius": {
+    "stops": [
+      // zoom is 0 and "rating" is 0 -> circle radius will be 0px
+      // zoom is 0 and "rating" is 5 -> circle radius will be 5px
+      [0, { "expression": [ "number_data", "rating" ] }]
+
+      // zoom is 20 and "rating" is 0 -> circle radius will be 4 * 0 = 0px
+      // zoom is 20 and "rating" is 5 -> circle radius will be 4 * 5 = 20px
+      [20, { "expression": [ "*", 4, ["number_data", "rating"] ] }]
+    ]
+  }
+}
+```
+
+
+## Property Expressions
+
+Property expressions are represented using a Lisp-like structured syntax tree.
+
+**Constants:**
+- `[ "ln2" ]`
+- `[ "pi" ]`
+- `[ "e" ]`
+
+**Literals:**
+- JSON string / number / boolean literal
+
+**Property lookup:**
+- Feature property:
+  - `[ "number_data", key_expr ]` reads `feature.properties[key_expr]`, coercing it to a number if necessary.
+  - `[ "string_data", key_expr ]` reads `feature.properties[key_expr]`, coercing it to a string if necessary.
+  - `[ "boolean_data", key_expr ]` reads `feature.properties[key_expr]`, coercing it to a boolean if necessary, with `0`, `''`, `null`, and missing properties mapping to `false`, and all other values mapping to `true`.
+  - `[ "has", key_expr ]` returns `true` if the property is present, false otherwise.
+  - `[ "typeof", key_expr ]` yields the data type of `feature.properties[key_expr]`: one of `'string'`, `'number'`, `'boolean'`, `'object'`, `'array'`, or, in the case that the property is not present, `'none'`.
+- `[ "geometry_type" ]` returns the value of `feature.geometry.type`.
+- `[ "string_id" ]`, `[ "number_id" ]` returns the value of `feature.id`.
+
+**Decision:**
+- `["if", boolean_expr, expr_if_true, expr_if_false]` 
+- `["switch", default_result_expr, [bool_expr1, result_expr1], [bool_expr2, result_expr2], ...]`
+- `["match", input_expr, default_result_expr, [test_expr1, result_expr1], [test_expr2, result_expr2]]`
+
+**Comparison and boolean operations:**
+- `[ "==", expr1, expr2]` (similar for `!=`)
+- `[ ">", lhs_expr, rhs_expr ]` (similar for <, >=, <=)
+- `[ "between", "[]"|"(]"|"[)"|"()", test_expr, left_boundary_expr, right_boundary_expr ]` - tests whether `test_expr` is "between" the left and right boundaries, with the first argument dictating whether the each boundary is exclusive or inclusive.
```

Comment:
Syntax proposal for intervals:

["intervals", _x_, _y<sub>1</sub>_, _n<sub>1</sub>_, ..., _y<sub>m</sub>_, _n<sub>m</sub>_, _y<sub>m+1</sub>_]

In this formulation, the input _x_ expression is followed by _m+1_ possible output _y_ expressions interspersed with _m_ breakpoints. In order to permit efficient evaluation, the breakpoints _n<sub>1</sub>_ < ... < _n<sub>m</sub>_ are required to be numeric literals in strictly ascending order; expressions are not allowed. The result is the first _y<sub>m_ such that _n<sub>m_ >= _x_, or _y<sub>m+1</sub>_ if _x_ is greater than all _n_'s.

Example:

```
"icon-image": {
    "type": "expression",
    "value": ["intervals", ["data", "population"],
        "small-dot", 10000, "medium-dot", 100000, "large-dot"]
}
```